### PR TITLE
Move that `maxMemoryCost` and `maxMemoryCountLimit` to config property. Add sync version API `diskImageDataExistsWithKey`

### PR DESCRIPTION
**Pull request by dreampiggy**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / reffers to the following issues: #1818 

### Pull Request Description

During SD 3.x -> 4.x, we remove that `diskImageExistsWithKey` sync version API and replaced with a async version API. This may because #781 deadlock

However, this change's pain is large than the gain. Sometimes we need to build something based on SDImageCache and need a sync version check. We have no choice to `force` to **simulate** a sync version API, see #1448.

For the deadlock, the reason is that previously, we expose our inner `ioQueue` to user, so if the user call `dispatch_sync` on the outside with the `ioQueue`, the deadlock occur.

But now, we do not expose that to the user, so this issue no longer happend. We should give back the sync version API.

And, I also copied some refactor code from 5.x branch. Including  `diskCacheWritingOptions` and replace `-[NSFileManager createFileAtPath]` with `-[NSData writeToURL:]`. For more information ,see #2148 

For thread-safe issue, see Apple Docs:

> The methods of the shared NSFileManager object can be called from multiple threads safely. However, if you use a delegate to receive notifications about the status of move, copy, remove, and link operations, you should create a unique instance of the file manager object, assign your delegate to that object, and use that file manager to initiate your operations.

### CODE REVIEW

**Comment by dreampiggy** [comment]
I wonder this `maxMemoryCost` & `maxMemoryCountLimit` change seems not really useful for current 4.x user. Maybe I shuold put this into 5.x branch :)

But the sync version API is really a good thing for some of user(including me) to use. I leave this here to allow more comment

**Comment by mythodeia** [comment]
@dreampiggy you can add this to v5 i think.

**Comment by dreampiggy** [comment]
@mythodeia It seems to be. Althrough I think we could release a minior version 4.3.x with some new APIs to make it easier to migrate to 5.x. But user seems have no strong demand to this...Althrough I have and I write one to `SDImageCache` category :)

**Comment by mythodeia** [comment]
@dreampiggy 
i agree to have maxMemoryCost & maxMemoryCountLimit inside SDImageCacheConfig.
SDImageCacheConfig should have all the configurations in one place. I think its better this way.
Most users do not have a need for this indeed. 
i use the v5 branch in a project of mine with all your commits and so far it looks good.
So you can add these in the v5 branch too.

**Comment by dreampiggy** [comment]
@mythodeia Actually the biggest change is in #2140 and not merged. But that PR changes too much, I'll seperate some of part into seperate PR to make it more clear.

**Comment by mythodeia** [comment]
👍

**Comment by dreampiggy** [comment]
# Current break change (Need change code)

### SDImageCache.h
+ `maxMemoryCost` MOVED to `SDImageCacheConfig.maxMemoryCost`
+ `maxMemoryCountLimit`, MOVED to `SDImageCacheConfig.maxMemoryCount`

File: SDWebImage/SDImageCache.m [hunk 1]
```diff
@@ -287,6 +289,21 @@ - (void)diskImageExistsWithKey:(nullable NSString *)key completion:(nullable SDW
     });
 }
 
+- (BOOL)diskImageDataExistsWithKey:(nullable NSString *)key {
```
Comment (dreampiggy) [inline SDWebImage/SDImageCache.m]: This seems no IO-safe(not be called on IO Queue). Maybe we should rename this method to `safeDiskImageDataExistsWithKey:`, and then sync/async public API call this method with `dispatch_sync(self.ioQueue)` and `dispatch_async(self.ioQueue)`

**Pull request merged**